application/src/test/java/org/thingsboard/server/service/mail/TbMailSenderTest.java

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright © 2016-2023 The Thingsboard Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.server.service.mail;
+
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
+import org.thingsboard.common.util.JacksonUtil;
+import javax.mail.MessagingException;
+import javax.mail.Session;
+import javax.mail.internet.MimeMessage;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+
+@RunWith(MockitoJUnitRunner.class)
+public class TbMailSenderTest {
+
+    @Test
+    public void testDoSendSendMail() {
+        ObjectNode jsonConfig = getBasicMailConfig();
+        TbMailSender tbMailSender = spy(new TbMailSender(mock(TbMailContextComponent.class), jsonConfig));
+
+        MimeMessage mimeMsg = new MimeMessage(Session.getInstance(new Properties()));
+        List<MimeMessage> mimeMessages = new ArrayList<>(1);
+        mimeMessages.add(mimeMsg);
+
+        Mockito.doNothing().when(tbMailSender).updateOauth2PasswordIfExpired();
```

Comment:
It is easier to use mock instead of spy. Spy is quite expensive to build and run 

You can call a real method in the mocked object like

`import static org.mockito.BDDMockito.willCallRealMethod;`
`willCallRealMethod().given(client).headersToMetaData(any(), any());`